dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/mapper/ProgramStageInstanceMapper.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.collections4.CollectionUtils.isNotEmpty;
+import static org.hisp.dhis.common.IdScheme.CODE;
+import static org.hisp.dhis.common.IdScheme.UID;
+import static org.hisp.dhis.event.EventStatus.fromInt;
+import static org.hisp.dhis.util.DateUtils.parseDate;
+
+import java.util.Date;
+import java.util.List;
+
+import org.hisp.dhis.dxf2.common.ImportOptions;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.dxf2.events.importer.context.WorkContext;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentitycomment.TrackedEntityComment;
+
+/**
+ * @author Luciano Fiandesio
+ */
+public class ProgramStageInstanceMapper extends AbstractMapper<Event, ProgramStageInstance>
+{
+    private final ProgramStageInstanceNoteMapper noteMapper;
+
+    public ProgramStageInstanceMapper( WorkContext ctx )
+    {
+        super( ctx );
+        noteMapper = new ProgramStageInstanceNoteMapper( ctx );
+    }
+
+    public ProgramStageInstance map( Event event )
+    {
+        ImportOptions importOptions = workContext.getImportOptions();
+        
+        ProgramStageInstance psi = new ProgramStageInstance();
+        
+        ProgramStageInstance programStageInstance = this.workContext.getProgramStageInstanceMap().get( event.getUid() );
+        if ( programStageInstance != null )
+        {
+            psi.setId( programStageInstance.getId() );
+        }
+
+        if ( importOptions.getIdSchemes().getProgramStageInstanceIdScheme().equals( CODE ) )
+        {
+            // TODO: Is this really correct?
+            psi.setCode( event.getUid() );
```

Comment:
No, that seems wrong.. code for tracker objects is anyways not properly implemented

Reply:
Ok, this was on of the questions I asked in the open points document: https://docs.google.com/document/d/1030nAuKxMSAYopNJj569inGgcJUS3DbqNrU-ZyZgSMI/edit#heading=h.p1u0vhrtwo4q